### ISSUE TITLE
Out of sequence chunks no longer considered a CorruptionErr and never deleted from disk

### ISSUE THREAD

**Issue by steveazz** [prometheus/prometheus#10043]
**What did you do?**

*Note:* This came from investigate a incident we had in production in https://gitlab.com/gitlab-com/gl-infra/production/-/issues/5998, we still don't know the root cause and are investigating in https://gitlab.com/gitlab-com/gl-infra/infrastructure/-/issues/14821 and when doing so we found the following:

1. Start seeing the following errors inside of Prometheus logs: `msg: Deletion of mmap chunk files failed, discarding chunk files completely err: cannot handle error: iterate on on-disk chunks: out of sequence m-mapped chunk for series ref 2667074520`.
2. We start looking at the stack trace where error is being return and saw the following:
    1. [`Head.Init`](https://github.com/prometheus/prometheus/blob/c92673fb1480cd6e3404a004522cbb851be4115d/tsdb/head.go#L474) is called
    2. [`loadMmappedChunks`](https://github.com/prometheus/prometheus/blob/c92673fb1480cd6e3404a004522cbb851be4115d/tsdb/head.go#L511) is called
    3. Finds out of sequence m-mapped chunk and [returns error](https://github.com/prometheus/prometheus/blob/c92673fb1480cd6e3404a004522cbb851be4115d/tsdb/head.go#L639)
    4. Check if error is of type [`CorruptionErr`](https://github.com/prometheus/prometheus/blob/c92673fb1480cd6e3404a004522cbb851be4115d/tsdb/head.go#L514)
        - Corruption counter never incremented.
        - Nothing else from the `loadMmappedChunks` return `CorruptionErr` so this check is either reduntant or `CorruptionErr` should be returned.
        - This was changed in https://github.com/prometheus/prometheus/pull/7229/files#diff-d9f058bd14a07ebec78d70aaa8d6e8e705bc09086e338b74b12666e7ac121605L542-R573
    6. Start recovery process by calling [`removeCorruptedMmappedChunks`](https://github.com/prometheus/prometheus/blob/c92673fb1480cd6e3404a004522cbb851be4115d/tsdb/head.go#L519)
    7. Deleting corrupted chunks with [`DeleteCorrupted`](https://github.com/prometheus/prometheus/blob/c92673fb1480cd6e3404a004522cbb851be4115d/tsdb/head.go#L670)
    8. [Check error type](https://github.com/prometheus/prometheus/blob/c92673fb1480cd6e3404a004522cbb851be4115d/tsdb/chunks/head_chunks.go#L767-L770)
        - In the past this used to be true for out of sequence, only after https://github.com/prometheus/prometheus/pull/7229 is merged.
        - This is where we get `cannot handle error` from the log message

**What did you expect to see?**

When [out of sequence m-mapped chunk](https://github.com/prometheus/prometheus/blob/c92673fb1480cd6e3404a004522cbb851be4115d/tsdb/head.go#L638-L640)
1. Increment the `prometheus_tsdb_mmap_chunk_corruptions_total` counter increments.
2. [`DeleteCorrupted` err assertion passes](https://github.com/prometheus/prometheus/blob/c92673fb1480cd6e3404a004522cbb851be4115d/tsdb/chunks/head_chunks.go#L767) 
3. [Delete chunk file from disk](https://github.com/prometheus/prometheus/blob/c92673fb1480cd6e3404a004522cbb851be4115d/tsdb/chunks/head_chunks.go#L772-L780)

**What did you see instead? Under which circumstances?**

We started seeing large [WAL files](https://gitlab.com/gitlab-com/gl-infra/production/-/issues/5998#note_750869286) and large `chunks_head` files which exhausted memory and disk usage. We aren't 100% sure that this specific "bug" resulted in resource exhaustion but, it still seems like it should delete the chunk from the disk, and might not helped recovery from a corrupted WAL/chunk

**Environment**

* System information:

```
$ uname -srm
Linux 5.4.120+ x86_64
```

Using the `quay.io/prometheus/prometheus` image in a GKE cluster.

* Prometheus version:

```
$ prometheus --version
prometheus, version 2.30.0 (branch: HEAD, revision: 37468d88dce85ac507f3fb7864c7d1c078e3e27d)
  build user:       root@e913c65aa170
  build date:       20210914-09:49:24
  go version:       go1.17.1
  platform:         linux/amd64
```

However this problem exists in `main` branch as well.

* Logs:
```
{"caller":"head.go:649","component":"tsdb","err":"cannot handle error: iterate on on-disk chunks: out of sequence m-mapped chunk for series ref 2981963907","level":"info","msg":"Deletion of mmap chunk files failed, discarding chunk files completely"}
```

Kudos to @mwasilew2 for helping me out

**Comment by steveazz** [comment]
I think the potential fix for this is https://github.com/steveazz/prometheus/commit/a6c3a836114d1289de285959793c82f579ee35b2 however, first I want to discuss **if** in https://github.com/prometheus/prometheus/pull/7229/files#diff-d9f058bd14a07ebec78d70aaa8d6e8e705bc09086e338b74b12666e7ac121605L542-R573 it was the intention to stop considering it a `CorruptionErr` and it's me misunderstanding the code.

**Comment by steveazz** [comment]
@codesome @bwplotka apologies for pinging you directly I appreciate that you folks are busy, but since this change happened in https://github.com/prometheus/prometheus/pull/7229 I'm curious about your opinion here and see if this is a bug or not.

Thank you so much for that hard work you put in Prometheus and Thanos 💪

**Comment by codesome** [comment]
Legit bug, we should be cleaning up the head chunk files. These out of sequence chunks are something mysterious (i.e. I haven't investigated the cause for it, though there were some other reason that could cause it which no longer errors and is fixed, but still something can cause it).

We cannot (or should't) delete it via the corruption error mechanism because we that is not technically a corruption but some other potential race. Since we are totally discarding the m-map chunks in this case, I suggest a "fix" of deleting all the m-map chunk files when we face this particular out of sequence error.

**Comment by codesome** [comment]
Okay, I take back my words about `We cannot (or should't) delete it via the corruption error mechanism`. Yes you are right, it was a bug introduced in #7229 and your suggested fix makes sense; that is the easiest way to delete all the m-map chunk files :)

**Comment by codesome** [comment]
We will need unit tests to avoid this bug from slipping in again.

**Comment by steveazz** [comment]
Thanks @codesome I agree that a test needs to be added to prevent this front happening again. 

I’d like to take this on if possible, but might be a bit slow because of holiday season!

**Comment by steveazz** [comment]
I'm now actively working on this again, I'm working on a test case where it can generate an out-of-sequence chunk so we can validate the chunk is deleted on disk. 

I'm struggling a little bit writing the test, but will reach out for help if I face a brick wall!

**Comment by codesome** [comment]
@steveazz you will find existing tests [here](https://github.com/prometheus/prometheus/blob/main/tsdb/chunks/head_chunks_test.go) where it's creating files and adding chunks manually.

**Comment by mwasilew2** [comment]
One other idea I had was to increase verbosity around this error a little bit. So in addition to fixing the logic that deletes the broken chunk (gets Prometheus out of the crashloop), the log message could contain more debug information. It currently contains a series ref which is a number. Example error message:

```
    cannot handle error: iterate on on-disk chunks: out of sequence m-mapped chunk for series ref 3691735024
```

and here's the error in the code: https://github.com/prometheus/prometheus/blob/main/tsdb/head.go#L630

I wonder if it would be possible to surface in the error message some more information about the series, for example the `__name__` label.

@steveazz feel free to include it in your PR. If on the other hand you think it would fit better in a separate PR let me know and I'm happy to open one.

**Comment by steveazz** [comment]
> One other idea I had was to increase verbosity around this error a little bit. So in addition to fixing the logic that deletes the broken chunk (gets Prometheus out of the crashloop), the log message could contain more debug information. It currently contains a series ref which is a number. Example error message:
> 
> ```
>     cannot handle error: iterate on on-disk chunks: out of sequence m-mapped chunk for series ref 3691735024
> ```
> 
> and here's the error in the code: https://github.com/prometheus/prometheus/blob/main/tsdb/head.go#L630
> 
> I wonder if it would be possible to surface in the error message some more information about the series, for example the `__name__` label.
> 
> @steveazz feel free to include it in your PR. If on the other hand you think it would fit better in a separate PR let me know and I'm happy to open one.

This actually got improved in https://github.com/prometheus/prometheus/commit/24827782cbc546adba99c7645dc8a01750023597  where it now shows the last chunk and the new chunk

**Comment by xvzf** [comment]
Hey folks

Is there an update on this? or any help we could provide? We just had this happening to one of our prometheus instances, filling up 80GB of disk-size with the retention set to max 64Gi

**Comment by codesome** [comment]
The issue title is a bit misleading. The files are actually deleted for out of sequence chunks. I am closing this issue because of that.

**Comment by codesome** [comment]
@xvzf if you have any specific details (logs, sizes of various TSDB artifacts, etc) and if it does not heal eventually (with a bit of increased disk size), please file another issue with details or comment the details on any relevant open issue.
